### PR TITLE
Milestone 3 patch for MDEV-21130

### PR DESCRIPTION
**Pull request by idoqo**
This is part of the ongoing project to switch the on-disk format of histograms to JSON [https://jira.mariadb.org/browse/MDEV-21130](MDEV-21130).

The goal of this milestone is to have a way to read a given JSON array and convert its elements to an in-memory data structure (currently a vector of strings).

This patch bundles the commits from Milestone 1 to 3 (not sure how to separate them as none of the milestone patches are ripe to be merged yet).

The relevant commit for this specific milestone is at https://github.com/MariaDB/server/commit/caf14c7455d444238679776b312749b7d3149084

### CODE REVIEW

**Review by spetrunia — CHANGES_REQUESTED** [review]

File: sql/sql_statistics.cc [hunk 1]
```diff
@@ -1580,6 +1594,140 @@ class Histogram_builder
...
+    test_parse_histogram_from_json();
+  }
+
+  static std::vector<std::string> parse_histogram_from_json(const char *json)
+  {
+    std::vector<std::string> hist_buckets= {};
+    enum json_types vt = json_get_array_items(json, json + strlen(json), hist_buckets);
+    printf("%d", vt);
+    printf("%zu", hist_buckets.size());
+
```
Comment (spetrunia) [inline sql/sql_statistics.cc:1661]: I've asked to print the diagnostics to stderr, where one can see it!  
So ,I have modified this patch to print to stderr, and also to print more:
https://gist.github.com/spetrunia/16f4a410b4702453f7ea7b2bb43abb7e

Then, I run this test:
```
create table ten(a int primary key);
insert into ten values (0),(1),(2),(3),(4),(5),(6),(7),(8),(9);
set histogram_type=json;
analyze table ten persistent for all;
```

and I see this output:
```
  ==============
  histogram: [
    "0",
    "1",
    "2",
    "3",
    "4",
    "5",
    "6",
    "7",
    "8",
    "9"
  ]
  json_get_array_items() returned -1
   JSV_BAD_JSON=-1, JSON_VALUE_ARRAY=2
  hist_buckets.size()=0
```

The histogram looks ok but the parsing didn't succeed?  Please fix this.

Comment (idoqo) [reply sql/sql_statistics.cc:1682]: Thanks for the patch! I tried running with `./mysql-test/mtr ...` but couldn't get the output (from `fprintf`) to show up. Do you mind sharing how you run it from your end?

Comment (spetrunia) [reply sql/sql_statistics.cc:1682]: mtr saves the server error log in the .err file.
```
cd mysql-test
./mtr statistics_json
vim var/log/mysqld.1.err
```

File: sql/sql_statistics.cc [hunk 2]
```diff
@@ -1651,83 +1664,81 @@ std::vector<std::string> bucket_bounds = {};
...
-   * are scalar values (i.e., strings, numbers, true or false), and returns JSV_BAD_JSON if:
-   * the original JSON is not an array OR the JSON array contains non-scalar elements.
-   */
-  static json_types json_get_array_items(const char *json, const char *json_end, std::vector<std::string> &container) {
-    json_engine_t je;
-    enum json_types value_type;
-    int vl;
-    const char *v;
+bool json_get_array_items(const char *json, const char *json_end, int *value_type, std::vector<std::string> &container) {
+  json_engine_t je;
```
Comment (spetrunia) [inline sql/sql_statistics.cc:1705]: I don't understand, what is the use of `value_type` for the caller?

Comment (idoqo) [reply sql/sql_statistics.cc:1705]: It's a way of telling the caller why parsing failed. That way, they know how the JSON structure was interpreted.

**Pull request closed without merging**